### PR TITLE
fix: Fix vertexing examples & update vertexing tutorial docs

### LINKED ISSUE #678: Broken vertexing examples
ACTS Version: Release 4.1 

Build Environment: lxplus with `CI/setup_cvmfs_lcg98.sh`

### AMVF

Command: 

```bash
ActsExampleAdaptiveMultiVertexFinder   \
  --events=10     \
  --rnd-seed=42   \
  --gen-cms-energy-gev=14000   \
  --gen-hard-process=Top:qqbar2ttbar=on   \
  --gen-npileup=140 \
  -l 0   \
  | tee log.py8_fly

```

Log: 

```
04:02:56    IPEstProp      ERROR     Step failed: EigenStepperError:3
04:02:56    AdaptiveMult   ERROR     Error in vertex finder: Step size adjustment exceeds maximum trials
04:10:38    AdaptiveMult   INFO      Found 1 vertices in event
04:10:38    AdaptiveMult   INFO      Found vertex at  0.0131516 0.00814755  0.0208274          0 with 104 tracks.
04:10:38    EventStore#5   VERBOSE   Added object 'protovertices'
04:10:38    Sequencer      INFO      finished event 5
04:11:19    AdaptiveMult   INFO      Found 1 vertices in event
04:11:19    AdaptiveMult   INFO      Found vertex at -0.00738011 -0.00728377   -0.721629           0 with 23 tracks.
04:11:19    EventStore#4   VERBOSE   Added object 'protovertices'
04:11:19    Sequencer      INFO      finished event 4
04:11:42    AdaptiveMult   INFO      Found 1 vertices in event
04:11:42    AdaptiveMult   INFO      Found vertex at  0.00258496 -0.00347993  -0.0779911           0 with 66 tracks.
04:11:42    EventStore#9   VERBOSE   Added object 'protovertices'
04:11:42    Sequencer      INFO      finished event 9
04:13:35    AdaptiveMult   INFO      Found 1 vertices in event
04:13:35    AdaptiveMult   INFO      Found vertex at  0.00862571 -0.00649257  -0.0660404           0 with 151 tracks.
04:13:35    EventStore#6   VERBOSE   Added object 'protovertices'
04:13:35    Sequencer      INFO      finished event 6
04:16:28    AdaptiveMult   INFO      Found 1 vertices in event
04:16:28    AdaptiveMult   INFO      Found vertex at -922.853 -7.50445 -1298.63        0 with 83 tracks.
04:16:28    EventStore#7   VERBOSE   Added object 'protovertices'
04:16:28    Sequencer      INFO      finished event 7
04:17:21    AdaptiveMult   INFO      Found 1 vertices in event
04:17:21    AdaptiveMult   INFO      Found vertex at -0.00422747  0.00582803   0.0399793           0 with 109 tracks.
04:17:21    EventStore#2   VERBOSE   Added object 'protovertices'
04:17:21    Sequencer      INFO      finished event 2
04:20:08    AdaptiveMult   INFO      Found 1 vertices in event
04:20:08    AdaptiveMult   INFO      Found vertex at  0.00843045 -0.00629033   -0.203474           0 with 69 tracks.
04:20:08    EventStore#3   VERBOSE   Added object 'protovertices'
04:20:08    Sequencer      INFO      finished event 3
terminate called after throwing an instance of 'std::runtime_error'
  what():  Failed to process event data
```

What I can see is: 

1. Step failed: EigenStepperError:3 

   This line prompts the error: https://github.com/acts-project/acts/blob/master/Core/include/Acts/Vertexing/ImpactPointEstimator.ipp#L163 

   The secDerivative turns out to be < 0, which should be > 0 

2. Error in vertex finder: Step size adjustment exceeds maximum trials

3. Not all primary vertices are reconstructed 
   Note I specify `--gen-npileup=140`, but only 1 vertex is reconstructed.

4. Not all events are processed 
   I want to process 10 events, but before it finishes, something goes wrong with `"std::runtime_error"`
   
 
### Reader Example 

The default [reader example]() uses IVF as the vertex finder. After I generate truth particle info with ((following [ACTS track tutorial](https://acts.readthedocs.io/en/latest/howto/run_fatras.html#generate-particle-truth-information): )

```bash
./ActsExamplePythia8 \
    --events=10     \
    --output-dir=data/gen/ttbar_mu140     \
    --output-csv=1     \
    --rnd-seed=42     \
    --gen-cms-energy-gev=14000     \
    --gen-hard-process=Top:qqbar2ttbar=on     \
    --gen-npileup=140
```

and then feed to the reader 

```bash
./ActsExampleVertexFinderReader --input-dir=data/gen/ttbar_mu140 
```

Log:

``` 
04:27:31    EventFilesRa   VERBOSE   Matching file event000000000-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000001-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000002-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000003-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000004-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000005-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000006-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000007-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000008-particles.csv
04:27:31    EventFilesRa   VERBOSE   Matching file event000000009-particles.csv
04:27:31    EventFilesRa   VERBOSE   Detected event range [0,9]
04:27:31    Sequencer      INFO      Added reader 'CsvParticleReader'
04:27:31    Sequencer      INFO      Added algorithm 'ParticleSelector'
04:27:31    Sequencer      INFO      Added algorithm 'ParticleSmearing'
04:27:31    Sequencer      INFO      Added algorithm 'TrackParametersPrinter'
04:27:31    Sequencer      INFO      Added algorithm 'IterativeVertexFinder'
04:27:31    Sequencer      INFO      Restrict requested number of events to available ones
04:27:31    Sequencer      INFO      Processing events [0, 10)
04:27:31    Sequencer      INFO      Starting event loop with 10 threads
04:27:31    Sequencer      INFO        0 services
04:27:31    Sequencer      INFO        0 context decorators
04:27:31    Sequencer      INFO        1 readers
04:27:31    Sequencer      INFO        4 algorithms
04:27:31    Sequencer      INFO        0 writers
04:27:31    TrackParamet   INFO      event 7 collection 'trackparameters' contains 11376 track parameters
04:27:31    TrackParamet   INFO      event 1 collection 'trackparameters' contains 9524 track parameters
04:27:31    TrackParamet   INFO      event 5 collection 'trackparameters' contains 8684 track parameters
04:27:31    TrackParamet   INFO      event 8 collection 'trackparameters' contains 9170 track parameters
04:27:31    TrackParamet   INFO      event 4 collection 'trackparameters' contains 9122 track parameters
04:27:31    TrackParamet   INFO      event 9 collection 'trackparameters' contains 9268 track parameters
04:27:31    TrackParamet   INFO      event 6 collection 'trackparameters' contains 10264 track parameters
04:27:31    TrackParamet   INFO      event 0 collection 'trackparameters' contains 10834 track parameters
04:27:31    TrackParamet   INFO      event 2 collection 'trackparameters' contains 11758 track parameters
04:27:31    TrackParamet   INFO      event 3 collection 'trackparameters' contains 12806 track parameters
04:27:31    TrackParamet   INFO        track 0
loc0:   -0.006793 +- 0.03         1.000
loc1:     0.02606 +- 0.03         0.000  1.000
phi:       0.1669 +- 0.01745      0.000  0.000  1.000
theta:      3.083 +- 0.01745      0.000  0.000  0.000  1.000
q/p:       0.1175 +- 0.0001176    0.000  0.000  0.000  0.000  1.000
time:       -1255 +- 1499         0.000  0.000  0.000  0.000  0.000  1.000
on surface undefined of type Acts::PerigeeSurface
04:27:31    TrackParamet   INFO        track 0
loc0:    -0.02674 +- 0.03         1.000
loc1:      -0.059 +- 0.03         0.000  1.000
phi:       -1.887 +- 0.01745      0.000  0.000  1.000
theta:      3.062 +- 0.01745      0.000  0.000  0.000  1.000
q/p:       0.3791 +- 0.0003792    0.000  0.000  0.000  0.000  1.000
time:        4790 +- 1499         0.000  0.000  0.000  0.000  0.000  1.000
on surface undefined of type Acts::PerigeeSurface
.....
04:28:16    TrackParamet   INFO        track 12805
loc0:     0.07135 +- 0.03         1.000
loc1:    0.008482 +- 0.03         0.000  1.000
phi:       -2.051 +- 0.01745      0.000  0.000  1.000
theta:      2.804 +- 0.01745      0.000  0.000  0.000  1.000
q/p:        -2.01 +- 0.002005     0.000  0.000  0.000  0.000  1.000
time:      -276.4 +- 1499         0.000  0.000  0.000  0.000  0.000  1.000
on surface undefined of type Acts::PerigeeSurface
04:28:16    IterativeVer   ERROR     Error in vertex finder: Numeric failure in calculation.
```
The last line returns with `Error in vertex finder: Numeric failure in calculation.` when calling IVF.

### PR DESCRIPTION
**Pull request by baschlag**
1. This PR fixes the vertexing examples issues we were seeing: primary vertexing specific track selection cuts that are needed for the fitters/finders to converge are added back (apparently they have been removed in some Examples rework in the past).
2. Also the vertexing tutorial docs were not up to date anymore with the ```TutorialVertexFinderAlgorithm``` in the Examples and are updated in this PR.

Closes #678

### CODE REVIEW

File: Examples/Run/Vertexing/Pythia8/AdaptiveMultiVertexFinderExample.cpp [hunk 1]
```diff
@@ -50,6 +50,9 @@ int main(int argc, char* argv[]) {
   selectParticles.outputParticles = "particles_selected";
   // smearing only works with charge particles for now
   selectParticles.removeNeutral = true;
+  selectParticles.absEtaMax = 2.5;
+  selectParticles.rhoMax = 4_mm;
+  selectParticles.ptMin = 500_MeV;
```
Comment (asalzburger) [inline Examples/Run/Vertexing/Pythia8/AdaptiveMultiVertexFinderExample.cpp:55]: Can we configure those with options?

File: Examples/Run/Vertexing/Pythia8/IterativeVertexFinderExample.cpp [hunk 2]
```diff
@@ -50,6 +50,9 @@ int main(int argc, char* argv[]) {
   selectParticles.outputParticles = "particles_selected";
   // smearing only works with charge particles for now
   selectParticles.removeNeutral = true;
+  selectParticles.absEtaMax = 2.5;
+  selectParticles.rhoMax = 4_mm;
+  selectParticles.ptMin = 500_MeV;
```
Comment (asalzburger) [inline Examples/Run/Vertexing/Pythia8/IterativeVertexFinderExample.cpp:55]: Same here ...

File: Examples/Run/Vertexing/Pythia8/TutorialVertexFinderExample.cpp [hunk 3]
```diff
@@ -50,6 +50,9 @@ int main(int argc, char* argv[]) {
   selectParticles.outputParticles = "particles_selected";
   // smearing only works with charge particles for now
   selectParticles.removeNeutral = true;
+  selectParticles.absEtaMax = 2.5;
+  selectParticles.rhoMax = 4_mm;
+  selectParticles.ptMin = 500_MeV;
```
Comment (asalzburger) [inline Examples/Run/Vertexing/Pythia8/TutorialVertexFinderExample.cpp:55]: Once again ...

File: Examples/Run/Vertexing/Pythia8/VertexFitterExample.cpp [hunk 4]
```diff
@@ -51,6 +51,9 @@ int main(int argc, char* argv[]) {
   selectParticles.outputParticles = "particles_selected";
   // smearing only works with charge particles for now
   selectParticles.removeNeutral = true;
+  selectParticles.absEtaMax = 2.5;
+  selectParticles.rhoMax = 4_mm;
+  selectParticles.ptMin = 500_MeV;
```
Comment (asalzburger) [inline Examples/Run/Vertexing/Pythia8/VertexFitterExample.cpp:56]: I keep repeating myself :-P

File: Examples/Run/Vertexing/VertexFinderReaderExample.cpp [hunk 5]
```diff
@@ -60,6 +60,9 @@ int main(int argc, char* argv[]) {
   selectParticles.outputParticles = "particles_selected";
   // smearing only works with charge particles for now
   selectParticles.removeNeutral = true;
+  selectParticles.absEtaMax = 2.5;
+  selectParticles.rhoMax = 4_mm;
+  selectParticles.ptMin = 500_MeV;
```
Comment (asalzburger) [inline Examples/Run/Vertexing/VertexFinderReaderExample.cpp:65]: ok ...

**Review by asalzburger — CHANGES_REQUESTED** [review]
In general this is fine, I was just wondering if we could have three parameters that are configurable with the ActsExample::Options ?

File: Examples/Run/Vertexing/Pythia8/AdaptiveMultiVertexFinderExample.cpp (hunk 1 above)
Comment (baschlag) [reply Examples/Run/Vertexing/Pythia8/AdaptiveMultiVertexFinderExample.cpp]: Done.

File: Examples/Run/Vertexing/Pythia8/IterativeVertexFinderExample.cpp (hunk 2 above)
Comment (baschlag) [reply Examples/Run/Vertexing/Pythia8/IterativeVertexFinderExample.cpp]: Done.

File: Examples/Run/Vertexing/Pythia8/TutorialVertexFinderExample.cpp (hunk 3 above)
Comment (baschlag) [reply Examples/Run/Vertexing/Pythia8/TutorialVertexFinderExample.cpp]: Once again done :)

File: Examples/Run/Vertexing/Pythia8/VertexFitterExample.cpp (hunk 4 above)
Comment (baschlag) [reply Examples/Run/Vertexing/Pythia8/VertexFitterExample.cpp]: Haha :-) they are all configurable now

**Comment by baschlag** [comment]
> In general this is fine, I was just wondering if we could have three parameters that are configurable with the ActsExample::Options ?

Good point, thanks. I've made these parameters configurable now.

**Review by asalzburger — CHANGES_REQUESTED** [review]
Almost, I see where you are coming from, but in all other Algorithm packages, we have one File such as e.g. `VertexingOptions` that then has the different `addIterativeFinderOptions(...)` etc. 

For keeping a homogeneous repo I think we should try to use the same approach everywhere.

**Comment by baschlag** [comment]
> Almost, I see where you are coming from, but in all other Algorithm packages, we have one File such as e.g. `VertexingOptions` that then has the different `addIterativeFinderOptions(...)` etc.
> 
> For keeping a homogeneous repo I think we should try to use the same approach everywhere.

Oh I missed that. Yes, that's a lot cleaner and avoids duplications. Perfect. Will change it.

**Comment by baschlag** [comment]
> Almost, I see where you are coming from, but in all other Algorithm packages, we have one File such as e.g. `VertexingOptions` that then has the different `addIterativeFinderOptions(...)` etc.
> 
> For keeping a homogeneous repo I think we should try to use the same approach everywhere.

Done. Please check.

**Review by asalzburger — APPROVED** [review]
Excellent.

**Comment by ZhangGang00** [comment]
Hi @baschlag ,

You required "vertexing-eta-max" and "vertexing-pt-min" to restrict the tracks in the barrel just for better track reconstruction. It does not mean ACTS AMVF cannot deal with tracks almost along Z axis, right? For FASER, our tracks are almost along Z axis. Is it OK for ACTS AMVF?

Cheers,
Gang

**Pull request merged**